File: exercises/ddd_application_layer/typescript/test/helpers/RecordingApplicationRepository.ts
import type { ApplicationRepository } from "../../src/domain/application/ApplicationRepository";
import type { Application } from "../../src/domain/application/Application";
import type { ApplicationId } from "../../src/domain/application/ApplicationId";

export class RecordingApplicationRepository implements ApplicationRepository {
  private readonly store = new Map<string, Application>();
  public readonly saved: Application[] = [];

  constructor(initial: Application[] = []) {
    for (const application of initial) {
      this.store.set(application.id.value, application);
    }
  }

  async getById(id: ApplicationId): Promise<Application | undefined> {
    return this.store.get(id.value);
  }

  async save(application: Application): Promise<void> {
    this.saved.push(application);
    this.store.set(application.id.value, application);
  }
}

File: exercises/ddd_application_layer/java/src/test/java/helpers/RecordingApplicationRepository.java
package helpers;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import domain.application.Application;
import domain.application.ApplicationId;
import domain.application.ApplicationRepository;

public class RecordingApplicationRepository implements ApplicationRepository {
    private Map<UUID, Application> store = new HashMap<>();
    public List<Application> saved = new ArrayList<>();

    public RecordingApplicationRepository() {
    }

    public RecordingApplicationRepository(Application initial) {
        this.store.put(initial.getId().getValue(), initial);
    }

    public RecordingApplicationRepository(List<Application> initial) {
        for (Application application : initial) {
            this.store.put(application.getId().getValue(), application);
        }
    }

    @Override
    public Application getById(ApplicationId id) {
        return this.store.get(id.getValue());
    }

    @Override
    public void save(Application application) {
        this.saved.add(application);
        this.store.put(application.getId().getValue(), application);
    }
}
